fix: improved usability of `tf-install.sh`

saforem2 opened the pull request:
## Example

```bash
#[2025-06-05 @ 15:25:52][~/p/s/textfox][🌱 saforem2/fix-install][✓]
; bash tf-install.sh
Path to firefox profile: '/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release'
Found profile at /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
This operation will copy the contents to your chrome dir, do you want to create a backup? (Y/N): y
Creating backup of the existing chrome directory in /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome_backup-20250605_152600
Do you want to install the user.js file? (Y/N): y
Installation completed, thank you for using textfox.
Execution time: 0h:00m:11s sec
```

## Copilot Summary

This pull request updates the `tf-install.sh` script to improve path handling, add error checks, and enhance user interaction during the installation process. The most important changes include normalizing and validating the Firefox profile path, improving the backup creation process, and refining the user.js installation flow.

### Path Handling Improvements:
* Expanded `~` to `$HOME`, removed trailing slashes, normalized double slashes, and stripped quotes from the `path` variable to ensure robust path handling. Added a check to verify the existence of the specified path and provide an error message if it does not exist.

### Backup Creation Enhancements:
* Improved the backup creation process for the `chrome` directory by generating a timestamped backup directory and providing user feedback during the operation.

### User.js Installation Flow:
* Enhanced the user.js installation logic by adding explicit feedback when skipping the installation and ensuring consistent handling of invalid inputs.

adriankarlen commented on tf-install.sh:
```diff
@@ -1,40 +1,51 @@
 #!/bin/bash
 
 read -e -p "Path to firefox profile: " path
-echo $path
+path="${path/#\~/$HOME}"  # Expand ~ to $HOME
+path="${path%/}"  # Remove trailing slash if exists
+path="$(echo "$path" | sed 's/\/\//\//g')"  # Normalize double slashes
+# remove single and double quotes
+path="$(echo "$path" | sed "s/[\"']//g")"
+if [ ! -d "$path" ]; then
+  echo "The specified path does not exist. Please check the path and try again."
+  exit 1
+fi
```
This could be a reprompting solution instead of exiting on path failure.

If you break out the path acquisition (with the improvements) to a function you could create a clean solution using a while loop.

saforem2 commented:
## Changes

So I made some decisions that (I think, hopefully) are logical and simplify things a bit but feel free to weigh in:

1. Allow passing the profile directory as an argument directly to `tf-install.sh`, e.g.:

    ```bash
    bash tf-install.sh <profile-dir>
    ```

    <details closed><summary>Example:</summary>

    ```bash
    #[2025-06-07 @ 19:33:31][~/p/s/textfox][🌱 saforem2/fix-install][📝🤷🏻‍♂️✓] [⏱️ 3s]
	; bash tf-install.sh "~/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release"
	   __            __  ____
	  / /____  _  __/ /_/ __/___  _  __
	 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
	/ /_/  __/>  </ /_/ __/ /_/ />  <
	\__/\___/_/|_|\__/_/  \____/_/|_|
	
	Installing textfox...
	Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
	[!!] Backing up existing chrome directory...
	/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193336.bak
	Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
	Do you want to install the user.js file? (Y/N): y
	user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
	✓ Installation completed
    ```

    </details>


1. If no arguments received, prompt user for input:

    <details closed><summary>Example:</summary>

    ```bash
	#[2025-06-07 @ 19:33:16][~/p/s/textfox][🌱 saforem2/fix-install][📝🤷🏻‍♂️✓]
	; bash tf-install.sh
	   __            __  ____
	  / /____  _  __/ /_/ __/___  _  __
	 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
	/ /_/  __/>  </ /_/ __/ /_/ />  <
	\__/\___/_/|_|\__/_/  \____/_/|_|
	
	Installing textfox...
	Path to Firefox profile: ~/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
	Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
	[!!] Backing up existing chrome directory...
	/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193319.bak
	Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
	Do you want to install the user.js file? (Y/N): y
	user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
	✓ Installation completed
	Execution time: 0h:00m:04s sec
    ```

    </details>


## Test Cases

<details closed><summary>Handle Different Cases</summary>

```bash
#[2025-06-07 @ 19:32:10][~/p/s/textfox][🌱 saforem2/fix-install][📝🤷🏻‍♂️✓]
; test_cases=("~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release" "'~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release'" "\"~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release\"" "~/Library//Application Support/Firefox/Profiles//3nvn2a07.default-release/" "~/broken_path" ) ; printf "%s\n" "${test_cases[@]}"
~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release
'~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release'
"~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release"
~/Library//Application Support/Firefox/Profiles//3nvn2a07.default-release/
~/broken_path

#[2025-06-07 @ 19:32:11][~/p/s/textfox][🌱 saforem2/fix-install][📝🤷🏻‍♂️✓]
; for tc in "${test_cases[@]}"; do echo "Testing:\n\t${tc}" && bash tf-install.sh "${tc}" ; done
Testing:
	~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release
   __            __  ____
  / /____  _  __/ /_/ __/___  _  __
 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
/ /_/  __/>  </ /_/ __/ /_/ />  <
\__/\___/_/|_|\__/_/  \____/_/|_|

Installing textfox...
Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
[!!] Backing up existing chrome directory...
/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193213.bak
Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
Do you want to install the user.js file? (Y/N): y
user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
✓ Installation completed
Testing:
	'~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release'
   __            __  ____
  / /____  _  __/ /_/ __/___  _  __
 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
/ /_/  __/>  </ /_/ __/ /_/ />  <
\__/\___/_/|_|\__/_/  \____/_/|_|

Installing textfox...
Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
[!!] Backing up existing chrome directory...
/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193214.bak
Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
Do you want to install the user.js file? (Y/N): y
user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
✓ Installation completed
Testing:
	"~/Library//Application Support/Firefox/Profiles/3nvn2a07.default-release"
   __            __  ____
  / /____  _  __/ /_/ __/___  _  __
 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
/ /_/  __/>  </ /_/ __/ /_/ />  <
\__/\___/_/|_|\__/_/  \____/_/|_|

Installing textfox...
Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
[!!] Backing up existing chrome directory...
/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193215.bak
Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
Do you want to install the user.js file? (Y/N): y
user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
✓ Installation completed
Testing:
	~/Library//Application Support/Firefox/Profiles//3nvn2a07.default-release/
   __            __  ____
  / /____  _  __/ /_/ __/___  _  __
 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
/ /_/  __/>  </ /_/ __/ /_/ />  <
\__/\___/_/|_|\__/_/  \____/_/|_|

Installing textfox...
Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
[!!] Backing up existing chrome directory...
/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193215.bak/chrome
Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
Do you want to install the user.js file? (Y/N): y
user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
✓ Installation completed
Testing:
	~/broken_path
   __            __  ____
  / /____  _  __/ /_/ __/___  _  __
 / __/ _ \| |/_/ __/ /_/ __ \| |/_/
/ /_/  __/>  </ /_/ __/ /_/ />  <
\__/\___/_/|_|\__/_/  \____/_/|_|

Installing textfox...
[!!] Directory /Users/samforeman/broken_path does not exist
Path to Firefox profile: ~/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
Using Firefox Profile @ /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release
[!!] Backing up existing chrome directory...
/Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome-20250607_193219.bak
Copying textfox/chrome/ -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/chrome/
Do you want to install the user.js file? (Y/N): y
user.js -> /Users/samforeman/Library/Application Support/Firefox/Profiles/3nvn2a07.default-release/user.js
✓ Installation completed
Execution time: 0h:00m:07s sec
```

</details>

---

## Copilot Summary

This pull request refactors the `tf-install.sh` script to improve its structure, readability, and functionality. The changes include modularizing the code into functions, adding input validation and path sanitization, and improving the user experience with better feedback and error handling.

### Code Refactoring and Modularity:
* Replaced the inline script logic with modular functions (`print_logo`, `clean_path`, `backup_existing_profile`, `copy_chrome`, `install_user_js`, `tf_install`) to improve readability and maintainability.

### Input Validation and Path Handling:
* Added the `clean_path` function to sanitize and normalize user-provided paths (e.g., expanding `~`, removing trailing slashes, and handling quotes). This ensures robust handling of input paths.

### Enhanced User Experience:
* Introduced a `print_logo` function to display a logo at the start of the script, improving the visual appeal of the installation process.
* Improved error handling and feedback by validating the existence of the Firefox profile directory and providing clear messages when paths are invalid or operations are skipped.

### Backup and Safety:
* Added the `backup_existing_profile` function to automatically back up the existing `chrome` directory with a timestamped name, replacing the previous manual backup prompt. ([tf-install.shL3-R89](diffhunk://#diff-522af4c763a120b2c860584caaa0c4c53201fa2b850c376413a212c931928ac7L3-R89)

adriankarlen reviewed (approved):
Looks good now, very nice improvements!
